core/src/main/java/org/apache/iceberg/BaseFileScanTask.java

```diff
@@ -20,71 +20,69 @@
 package org.apache.iceberg;
 
 import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
 import java.util.Iterator;
 import java.util.List;
 import java.util.NoSuchElementException;
+import javax.annotation.Nullable;
 import org.apache.iceberg.expressions.Expression;
 import org.apache.iceberg.expressions.ResidualEvaluator;
 import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
 import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
 import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
-
-class BaseFileScanTask implements FileScanTask {
-  private final DataFile file;
-  private final DeleteFile[] deletes;
-  private final String schemaString;
-  private final String specString;
-  private final ResidualEvaluator residuals;
-
-  private transient PartitionSpec spec = null;
-
-  BaseFileScanTask(DataFile file, DeleteFile[] deletes, String schemaString, String specString,
-                   ResidualEvaluator residuals) {
-    this.file = file;
-    this.deletes = deletes != null ? deletes : new DeleteFile[0];
-    this.schemaString = schemaString;
-    this.specString = specString;
-    this.residuals = residuals;
+import org.immutables.value.Value;
+
+@Value.Immutable
+abstract class BaseFileScanTask implements FileScanTask {
```

Comment:
if we want to make this JSON serializable/deserialize, all we have to do is
```
@JsonSerialize(as = ImmutableBaseFileScanTask.class)
@JsonDeserialize(as = ImmutableBaseFileScanTask.class)
```

I have omitted this for now until we really need it.

Reply:
Unfortunately we cannot use @JsonSerialize and @JsonDeserialize in Trino and PrestoDB, see https://github.com/trinodb/trino/pull/8754#discussion_r681375336 for context.

Reply:
That's why we are actually experimenting with serializing everything as a JSON string, just use @JsonValue to store the values, and have a `FileScanTaskParser` that creates the object directly from the JSON string. Please let me know if you have any better ways to achieve that.

Reply:
I do not now how Trino&Iceberg integration looks like exactly, but maybe it's worth checking whether the generated immutable class can offer something that could be used? The purpose of the `@JsonSerialize` / `@JsonDeserialize` annotations is to generate a `@JsonCreator` factory method and adding `@JsonProperty` annotations on the generated immutable class. See https://immutables.github.io/json.html for some details.

Reply:
@jackye1995 did you have a change checking up on my previous comment?

Reply:
Not sure that I agree that 'serializing everything as a JSON string' is the right way forward. Nor do I think that we should forgo making our code safer, cleaner and easier to use because an engine cant _directly_ handle it.

As @nastra pointed out this annotation generates the annotations and objects that Trino needs to serialize (namely `@JsonCreator` and `@JsonProperty`.